modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/IgniteSqlImpl.java

```diff
@@ -54,7 +82,7 @@ public Session createSession() {
     /** {@inheritDoc} */
     @Override
     public SessionBuilder sessionBuilder() {
-        return new SessionBuilderImpl(qryProc, transactions, new HashMap<>());
+        return new SessionBuilderImpl(sessions, qryProc, transactions, new HashMap<>());
```

Comment:
You pass sessions collection to SessionBuilderImpl, which leaks to the user side.
SessionBuilderImpl has no locks/flags that prevents putting data into collections, when node is stopping. After executor will be stopped, noone will cleanup the collection.

Maybe, we can pass a listener interface with `onSessionCreate()` method and do all stuff outside the builder?

Reply:
now it's guarded with busyLock